nomad/state/state_store.go

```diff
@@ -5709,6 +5684,44 @@ func (s *StateStore) UpsertEvents(index uint64, events *structs.Events) error {
 	return txn.Commit()
 }
 
+func insertAndPruneEvents(txn *txn, keep int, events *structs.Events) error {
+	// Insert new events
+	if err := txn.Insert("events", events); err != nil {
+		return err
+	}
+
+	// Reverse iterator starting from newest to oldest
+	iter, err := txn.GetReverse("events", "id")
+	if err != nil {
+		return fmt.Errorf("events lookup failed: %v", err)
+	}
+
+	var count int
+	var oldEvents []*structs.Events
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+		e := raw.(*structs.Events)
+
+		count += len(e.Events)
+
+		// Count is greater than amount to keep, mark for deletion
+		if count > keep {
+			oldEvents = append(oldEvents, e)
+		}
+	}
+
+	for _, e := range oldEvents {
```

Comment:
This looks like we end up needing to iterate over the entire event stream for every insert? (Actually, `(new)+2(old)`.) In the common case after an initial warmup and with a full event buffer, I think we'll be doing that iteration to delete a single event per insert.

Given that the events are immutable I think we can use their index as a counter. Is it at all possible to use something like `GetReverse` or `ReverseLowerBound` to iterate from the end of the events and get the appropriate count of events to delete?

Reply:
Events are immutable, but for each entry, there could be 1-n events (large batch operation), so I'm not sure if we can rely on index as a counter. If there was a way to check if the buffer is passed the initial warmup we'd be able to just remove the amount we are inserting from the top

Reply:
Ah. And that prevents us from doing a `DeletePrefix` for an index as well.

Is there a way we could track the total count, even if its only in-memory on the leader and unpopulated until we do the first event insert? This feels like it could be a premature optimization but I expect we'll have a very large set of events to iterate over (the default is 100 though?), and it's not like we're just bumping a pointer over a slice to count -- we have to dereference as well to get the events to count them.

Maybe the expected number of events is low enough that it's just few hundred pointer bumps and dereferences, in which case it probably doesn't matter?

Reply:
100 is the default currently, but once we get enough testing and are able to gauge a range in size for how big 100 or 1000 events really are it could get larger.. I'm trying to think through a few options for how we could accomplish this because iterating the entire set on every insert on the fast path of the FSM isn't great..

The state store could have an atomic counter we use to track? :thinking:

Reply:
Hm, so memdb tracks `Events` but the limit is based on the count of `Event` structs? That seems potentially problematic as our batched raft messages can be > 100. Consul and Vault token revocation each do batches of 1000, which if I'm reading this properly will delete every message in statestore leaving it empty. I'm also concerned the O(n) insertion cost will be untenable for busy clusters who assume a larger buffer only incurs extra memory+storage.

I know it's probably painful but can we change statestore to store `Event`s directly? They have the index so starting from a particular index is still possible (although you have to keep iterating in reverse until you reach index-1 which is a bit of extra logic).

I think that would allow for the intuitive O(1) insertion cost (add 1 event at a time, pop 1 event at a time), although operating on a per-Event basis instead of batches may incur *other* costs. :thinking:

Reply:
The other alternative is keeping `Events` as the unit of storage, but also using it as the unit for limits. That might be ok as our batches are of things that are relatively small? Although our additional materialization part might ruin that.